Profile

jaha01 opened the pull request:
Profile VC was rewrote to VIP, CarsModel aslo, was realized CarsService
P.S. i delete pods again...

jaha01 committed:
first commit

jaha01 committed:
Profile VC was rewrote to VIP

Profile VC was rewrote to VIP, CarsModel aslo, was realized CarsService
P.S. i delete pods again...

jaha01 committed:
saving data

Get and save profile info from/to firebase

jaha01 committed:
Merge branch 'main' into Profile

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelInteractor.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    // MARK: - Private properties
+    var presenter: CarsModelPresenter!
+    private let carsService: CarsService!
```
Из каких соображений ты поставил тут анврап? Вариант "потому что так у презентера выше" не принимается)

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    // MARK: - Private properties
+    var interactor: CarsModelInteractor!
+    private let searchVC = UISearchController(searchResultsController: nil)
+    private let list = [String]()
```
Список чего?

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,76 @@
+
+        view.addSubview(tableView)
+        navigationItem.title = "Cars List"
+        tableView.delegate = self
```
Если table view объявлять как lazy var, то настройку делегата и датасорса можно класть туда же, внутрь создающей кложуры.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        cell.textLabel?.text = "BMW"
```
Я ж так понимаю, это хардкод, который будет выпилен в следующем реквесте?

YarMag commented on autoService/Profile/Cars Info/CarsService.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    func loadCars(query: String, completion: @escaping(Result<[String], Error>)->Void) {
+        // можно ли так делать??? с path
+        networkClient.request(path: "https://qr2.mydigimenu.com/36f62b7c-e406-48f5-870f-02512e313c7a/menu-page?menuID=323", method: "POST", query: query) { (result: Result<CarsSuggestions, Error>) in
```
Нет. Вспомни, мы обсуждали, что у тебя есть хост (https://qr2.mydigimenu.com) и path. Причем путь у тебя тут какой-то подозрительный - ты пишешь в методе POST, но при этом в строке урла все равно есть аргументы.

YarMag replied:
Самая близкая аналогия, которая тут приходит в голову - представь, что ты хочешь сходить к кому-то в гости и спрашиваешь у него адрес. Хост - это, условно, адрес дома. А путь - конкретная квартира, в которую ты хочешь прийти. Дом один, а квартир много. Так и тут - хост один, а путей (урлов, или методов сервера, названий много, суть одна) относительно этого хоста - сильно больше.

YarMag commented on autoService/Profile/Cars Info/CarsService.swift:
```diff
@@ -0,0 +1,37 @@
+            case .success(let data) :
+                print("data = \(data)")
+                //completion(.success(data))
+                // map. -> convert to array
```
С map проблемы возникли?

YarMag commented on autoService/Profile/Cars Info/NetwokClient.swift:
```diff
@@ -0,0 +1,80 @@
+import Foundation
+
+class NetworkClient : NSObject {
+    private var networkConfiguration = NetworkConfig() // init
```
А где, собственно, init?

YarMag commented on autoService/Profile/Cars Info/NetwokClient.swift:
```diff
@@ -0,0 +1,80 @@
+            return
+        }
+        var urlRequest = URLRequest(url: url)
+        let headers = networkConfiguration.headers()
```
Объявляй переменные максимально близко к месту их использования. Header и его использование у тебя разрываются работой с jsonData, не надо так. Группируй блоки кода по смыслу/функциональности.

YarMag commented on autoService/Profile/Cars Info/NetwokClient.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+        var urlRequest = URLRequest(url: url)
+        let headers = networkConfiguration.headers()
+        let json: [String: Any] = ["query": query as Any]
```
Метод получается неуниверсальный. Тебе надо передавать целиком Codable-объект, перегонять его в JSON как есть и класть результат в httpBody. Сейчас ты завязан на то, что у тебя в body должен лежать объект с одним ключом query и каким-то содержимым.

YarMag commented on autoService/Profile/Cars Info/NetwokClient.swift:
```diff
@@ -0,0 +1,80 @@
+                case 200..<400:
+                    let json = String.init(data: data, encoding: .utf8)
+                    print("json = \(json)")
+                    self.jsonDecoder.keyDecodingStrategy = .convertFromSnakeCase
```
Вот это штука хорошая, но тогда тебе надо поменять имена в модели, с которой ты работаешь. Ты либо пишешь структуру, у которой есть поля вида field_name, либо пишешь fieldName и используешь .convertFromSnakeCase.

YarMag commented on autoService/Profile/Cars Info/URLComponent.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+struct URLComponent {
+    static let path = "/suggestions/api/4_1/rs/suggest/car_brand"
```
Ммм, это не должно быть экстеншном у URLComponent. Представь, что у тебя появится несколько урлов, к которым нужно ходить. Как тогда выкручиваться будешь? Я бы предложил, скорее, enum с ассоциированными строковыми значениями. По-моему, на курсах вам как раз такой подход и показывали.

YarMag commented on autoService/Profile/ProfileInfo.swift:
```diff
@@ -0,0 +1,39 @@
+    let item: [Object]
+}
+
+struct Object {
```
Имя структуры неинформативное. Непонятно, что за объект.

YarMag commented on autoService/Profile/ProfileInfo.swift:
```diff
@@ -0,0 +1,39 @@
+//    case car(value: String, placeholder: String)
+}
+
+struct ProfileInfoNew {
```
Вот так лучше не делать. Если в проекте начинают появляться суффиксы вида Old, New, Refactored, 1, 2 - самое время брать метлу и наводить порядок.

YarMag commented on autoService/Profile/ProfileInfo.swift:
```diff
@@ -0,0 +1,39 @@
+
+struct ProfileInfo {
+    let section: String
+    let item: [Object]
```
Дезинформирующее имя у поля. Название item (единственное число), а тип поля - массив структур (т.е. множественное число). Такие вещи всегда должны быть синхронизированы.

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -13,9 +13,31 @@ final class ProfileInteractor {
+    func onViewDidLoad() {
+        dbService.setupProfileInfoListeners { info in
+            let ProfileDetails =
+            ProfileInfoNew.init(email: info["email"] as! String, phone: info["phone"] as! String, name: info["name"] as! String, city: info["city"] as! String, experience: info["experience"] as! String, bday: info["bday"] as! String, brand: info["brand"] as! String, model: info["model"] as! String, year: info["year"] as! String, volume: info["volume"] as! String, mileage: info["mileage"] as! Stri ...
```
Если у тебя при написании возникло ощущение, что что-то не так - так и есть. Представь, что этот метод будет вызываться на нескольких экранах. Тогда каждый раз придется писать вот такую портянку кода. Подобные вещи лучше спрятать внутри dbService, тогда каждый класс-пользователь будет получать сразу красивую, распарсенную структуру, а все неприглядные потроха будут написаны только в одном месте и скрыты от лишних глаз.

YarMag commented on autoService/Profile/ProfileInfo.swift:
```diff
@@ -0,0 +1,39 @@
+    let name: String
+    let city: String
+    let experience: String
+    let bday: String
```
Bad day? Big day? Bomb day? Лучше лишний раз не пользоваться сокращениями, по контексту может быть не всегда понятно, о чем речь.

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -13,9 +13,31 @@ final class ProfileInteractor {
+            let ProfileDetails =
+            ProfileInfoNew.init(email: info["email"] as! String, phone: info["phone"] as! String, name: info["name"] as! String, city: info["city"] as! String, experience: info["experience"] as! String, bday: info["bday"] as! String, brand: info["brand"] as! String, model: info["model"] as! String, year: info["year"] as! String, volume: info["volume"] as! String, mileage: info["mileage"] as! Stri ...
+
+            self.presenter.prepareToShowProfileData(ProfileDetails)
```
А нет ли у тебя тут сильного захвата?

jaha01 replied:
захвата чего?

YarMag replied:
retain cycle

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -13,9 +13,31 @@ final class ProfileInteractor {
+    
+    func onViewDidLoad() {
+        dbService.setupProfileInfoListeners { info in
+            let ProfileDetails =
```
С большой буквы мы называем разве что типы данных. Переменные - никогда.

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -13,9 +13,31 @@ final class ProfileInteractor {
+        router.goToCarsModel()
+    }
+    
+    func updatePofileInfo(info: ProfileInfoNew) {
```
Profile, буква пропущена.

YarMag commented on autoService/Profile/ProfileTableHeader.swift:
```diff
@@ -0,0 +1,48 @@
+
+    // MARK: - Private properties
+    
+    var photo: UIImageView = {
```
Почему var, еще и публичный?

YarMag commented on autoService/Profile/ProfileTableHeader.swift:
```diff
@@ -0,0 +1,48 @@
+        return image
+    }()
+
+    override init(frame: CGRect) {
```
MARK на инит бы не помешал.

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -16,6 +16,7 @@ final class DBService {
     private var allItems = [JournalItem]()
     private let authService: AuthService
     private let journalHistoryItems = "journalHistoryItems"
+    private let profileInformation = "ProfileInformation"
```
Старайся выдерживать имена в одном стиле. journalHistoryItems у тебя с маленькой буквы, а ProfileInformation с большой. Почему?

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -45,7 +46,24 @@ final class DBService {
         ref.child(authService.getUserID()).child(journalHistoryItems).child(id).removeValue()
     }
+    
+    func setupProfileInfoListeners(handler: @escaping ([String: Any]) -> Void) {
```
Лучше не [String: Any] возвращать, а человеческий массив объектов, с которыми будет работать тот, кто этим пользуется. Ты если идешь к врачу, он тебе не справочники дает со словами "почитайте вот на этих и на этих страницах, какие лекарства нужны в вашем случае", а сразу выдает готовый рецепт для аптеки. Тут идея ровно та же самая.

jaha01 replied:
Готово, теперб все спрятано.

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -45,7 +46,24 @@ final class DBService {
+    func uploadProfileInfo(profileInfo: ProfileInfoNew) {
+        
+        let parent = ref.child(authService.getUserID()).child(profileInformation)
+        let values = ["id": parent.key!, "email": profileInfo.email, "phone": profileInfo.phone, "name": profileInfo.name, "city": profileInfo.city, "experience": profileInfo.experience, "bday": profileInfo.bday,"brand": profileInfo.brand, "model": profileInfo.model, "year": profileInfo.year, "volume": profileInfo.volume, "mileage": profileInfo.mileage] as [String: Any]
```
А вот тут ты как раз сделал правильно - ты спрятал подготовку данных для Firebase внутри сервиса, и об этом больше никто не знает. 
Обрати еще внимание, к вопросу о том, что все должно работать единообразно. Декодирование из объекта Firebase в человеческий тип у тебя происходит за пределами DBService. Кодирование из человеческого типа в объект Firebase - внутри DBService. Такие моменты должны настораживать, когда одни и те же (с концептуальной точки зрения) операции находятся в сильно разных местах.

Можешь просто зарезолвить коммент, как прочтешь.

jaha01 replied:
Готово, теперь все спрятано.

jaha01 committed:
getting cars list - success

Now our service successfully gets cars list.
Also some tabulations were made.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelInteractor.swift:
```diff
@@ -0,0 +1,31 @@
+
+final class CarsModelInteractor {
+    
+    // MARK: - Private properties
```
Давай договоримся, что перед // MARK: и после него будем оставлять пустую строку. Когда все слитно, оно смотрится не очень.

YarMag replied:
Кстати, сразу не заметил - тут у тебя тоже секция в MARK называется Private properties, а внутри спрятался публичный презентер. Не надо так. Вообще, смотрю, ты редко где ставишь секцию Public properties.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelPresenter.swift:
```diff
@@ -0,0 +1,20 @@
+        controller.pushCarsList(cars: cars)
+    }
+    
+    func showErrorGetCarsList(error: String) {
```
Лучше просто showError. Т.к. метод у тебя сейчас универсальный, он умеет показывать любую ошибку, не только по получению списка машин.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+final class CarsModelViewController: UIViewController {
+    
+    // MARK: - Private properties
+    var interactor: CarsModelInteractor!
```
Так интерактор же очень даже публичный. И это тот редкий случай, когда это нормально.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        return table
+    }()
+    
+    // MARK: - Public methods
```
Вот сравни форматирование вот этого куска с Public Methods и того, что ниже, с Private Methods. Согласись, что такой вариант чуть поприятнее.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
```
Немного странный метод. У таблицы есть свойство allowsSelection, его можно выставить в false. Тогда селектор не будет работать вообще. Можно у самой ячейки в методе cellForRow проставить cell.selectionStyle = .none (или как-то так), тогда у тебя ячейка будет селектиться, но бэкграунд останется прежним.

jaha01 replied:
Думал этот метод использовать для выбора автомобиля, после клика информация передалась бы в ProfileViewController

YarMag replied:
Ага, понял. Но deselect тебе все равно не нужен, достаточно cell.selectionStyle = .none в методе конфигурации ячейки.

YarMag commented on autoService/Services/CarsService.swift:
```diff
@@ -0,0 +1,36 @@
+            switch result {
+            case .success(let data) :
+                let brands = data.suggestions.map({ $0.data.nameRu
+                })
```
Скобка странно стоит. Ctrl + I не забывай делать.

jaha01 replied:
во всем файле сделано уже

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -45,7 +46,25 @@ final class DBService {
+        ref.child(authService.getUserID()).child(profileInformation).observe(.value) { snapshot in
+            if let dictionary = snapshot.value as? [String: Any] {
+                let profileDetails =
+                ProfileInfo.init(email: dictionary["email"] as! String, phone: dictionary["phone"] as! String, name: dictionary["name"] as! String, city: dictionary["city"] as! String, experience: dictionary["experience"] as! String, birthday: dictionary["bday"] as! String, brand: dictionary["brand"] as! String, model: dictionary["model"] as! String, year: dictionary["year"] as! String, volume: d ...
```
Все равно выглядит стремно. Давай лучше сделаем у ProfileInfo кастомный инит, вида

```swift
init(json: [String: Any]) {
    // инициализация полей из ключей json-а
}
```

Просто прикинь, что у тебя ProfileInfo будет использоваться в разных местах приложения и тебе каждый раз придется писать такую портянку кода, чтобы просто создать структуру. Такие вещи как раз лучше в инициализатор прятать. Заодно тебе намного проще будет редактировать код, если вдруг у тебя в json появится еще одно поле, которое ты хочешь вытащить. Что проще - поменять код в одном init, или редачить вызовы инита во всем приложении, где в каждом таком ините по 15 параметров передается?

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -45,7 +46,25 @@ final class DBService {
+    func uploadProfileInfo(profileInfo: ProfileInfo) {
+        
+        let parent = ref.child(authService.getUserID()).child(profileInformation)
+        let values = ["id": parent.key!, "email": profileInfo.email, "phone": profileInfo.phone, "name": profileInfo.name, "city": profileInfo.city, "experience": profileInfo.experience, "bday": profileInfo.birthday,"brand": profileInfo.brand, "model": profileInfo.model, "year": profileInfo.year, "volume": profileInfo.volume, "mileage": profileInfo.mileage] as [String: Any]
```
Вот это, кстати, тоже можно затолкать внутрь ProfileInfo и сделать у него метод toJson() -> [String: Any]. Помнишь первый принцип ООП - инкапсуляцию? Вот она как раз в том числе про то, чтобы группировать вместе связанный код, чтобы ничего лишнего наружу не торчало.

YarMag commented on autoService/Services/Requests.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+enum Requests {
+    case common
```
Почему выбрал такое имя?

jaha01 replied:
общий путь...

YarMag replied:
Общий для кого? Почему он общий? Им никто пользоваться в общем порядке не обязан. Это просто путь, который нужен для получения списка машин. Завтра у тебя добавится еще три урла - будешь переименовывать?) Название должно четко отражать назначение той или иной сущности. Чем более абстрактное имя, тем более размыта ответственность. А тут у тебя ответственность вполне конкретная, значит, и имя должно быть соответствующее.

YarMag commented on autoService/Services/NetworkService.swift:
```diff
@@ -0,0 +1,70 @@
+    }()
+    
+    // MARK: - Public methods
+    func request<T:Codable>(path: String, method: String? = nil, body: Codable? = nil, completion: @escaping(Result<T,Error>)->Void) {
```
Знаешь, как еще делают, чтоб прям совсем по красоте было? Метод можно передавать строкой, но тогда никто не мешает тебе в качестве параметра прокинуть что-то совсем левое. А можно объявить

```swift
enum HttpMethod: String {
    case get = "GET"
    case post = "POST"
    case patch = "PATCH"
    case PUT = "PUT"
    case delete = "DELETE"
}
```

Тогда ты сможешь передавать в качестве аргумента один из кейсов енама, т.е. у тебя аргумент становится method: HttpMethod. И его можно сделать не опциональным, а с дефолтным значением method: HttpMethod = .get. И ниже ты будешь присваивать это как urlRequest.httpMethod = method.rawValue, без всяких проверок в if, они становятся не нужны.

YarMag commented on autoService/Services/Requests.swift:
```diff
@@ -0,0 +1,24 @@
+}
+
+
+struct CarsListRequestBody: Codable {
```
А вот этот товарищ вообще отдельного файла заслуживает. Во всяком случае, его нахождение в файле Requests.swift точно не интуитивно.

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -11,11 +11,29 @@ final class ProfileInteractor {
+    
+    // MARK: - Public methods
+    func onViewDidLoad() {
+        dbService.setupProfileInfoListeners { profileInfo in
```
Почему нет [weak self]? Утечкой пахнет, однако.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+    
+    private func setConstraints() {
+        
```
Вот эти пустые пространства внутри метода странно выглядят. Они ж ничего не отделяют, у тебя весь метод состоит из одной-единственной операции на активацию констрейнтов.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    func pushCarsList(cars: [String]) {
+        carsList = cars
+        DispatchQueue.main.async {
+            self.tableView.reloadData()
```
Что-то ты недолюбливаешь [weak self] в этот раз)

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        
+    }
+    
+    func pushCarsList(cars: [String]) {
```
push не совсем подходящее слово. Обычно, когда пишут push, подразумевают какую-то операцию добавления в конец некоторого массива. Думаю, тут достаточно showCarsList.

YarMag commented on autoService/Services/CarsSuggestions.swift:
```diff
@@ -0,0 +1,24 @@
+    let suggestions: [CarsList]
+}
+
+struct CarsList: Codable {
```
Так, что-то я тут с неймингом прощелкал. Ты назвал тип CarsList - список машин. Но структура по факту представляет один-единственный элемент этого списка. 

Я бы предложил назвать это CarInfo (Data в Swift все-таки чаще используется при работе с бинарными данными, байтами). А тот тип, что ниже - CarNameInfo.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelPresenter.swift:
```diff
@@ -0,0 +1,20 @@
+    weak var controller: CarsModelViewController!
+    
+    func prepareCarsList(cars: [String]) {
+        controller.pushCarsList(cars: cars)
```
У меня начало появляться подозрение, что я тебе не совсем правильно посоветовал. Вообще презентер - это как раз идеальное место, чтобы делать тут маппинг для UI и отсекать все лишнее. Т.е. то, что сейчас делается в CarsService, по-хорошему, должно делаться здесь. Тут мой косяк. У нас презентер сейчас вообще непонятно, зачем нужен.

Давай поступим следующим образом. CarsService в интерактор пусть возвращает структуру CarsSuggestions как есть, без всяких маппингов. Интерактор в презентер пусть отдает [CarInfo]. Т.е. не первоначальный объект CarsSuggestions, а сами саджесты, которые внутри него (ибо презентеру, по сути, как раз массив и нужен). А вот презентер уже на своем уровне отсекает от CarInfo все лишнее и оставляет только самый минимум, нужный для представления, т.е. массив строк. Это как раз его задача. А то мы что-то увлеклись и вообще презентеру кода не оставили.

YarMag replied:
Если что - переспроси в личке.

jaha01 committed:
successfully get cars brand

load cars brand and use for this bottom sheet, but it's not done yet, maybe 5% left

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelInteractor.swift:
```diff
@@ -0,0 +1,36 @@
+final class CarsModelInteractor {
+    
+    // MARK: - Public properties
+    var presenter: CarsModelPresenter!
```
Перед и после MARK по одной пустой строке. Перед марком вижу, после марка надо еще добавить.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelInteractor.swift:
```diff
@@ -0,0 +1,36 @@
+
+    private let carsService: CarsService
+    
+    // MARK: - Public methods
```
Инициализаторы обычно заслуживают собственных MARK-ов. Т.е. должны быть // MARK: - Initializers и // MARK: - Public methods. Во всех стайл гайдах такое практикуют. 

Рекомендую (если до сих пор я этого еще не сделал) поизучать вот это - https://github.com/kodecocodes/swift-style-guide . Обрати внимание - там про марки внутри класса ничего не сказано. Там указано, как обозначать, если марк над экстеншном. Поэтому давай зафиксируем такой формат:


```swift
class MyClass {
    
    // MARK: - Initializer, ставим пустую строку перед и после марка.
    
    init() { 
        // do something
    }
}

// MARK: - MyProtocol implementation, перед марком пустая строка есть, после не нужно
extension MyClass: MyProtocol {
    // MyProtocol methods
}
```

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelPresenter.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+final class CarsModelPresenter {
```
MARK-ов нет. Привыкай их расставлять на автомате, на работе все равно заставят)

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelPresenter.swift:
```diff
@@ -0,0 +1,23 @@
+    func prepareCarsList(cars: [CarInfo]) {
+        
+        let brands = cars.map { $0.data.nameRu
+        }
```
Странное расположение закрывающей скобки. Ей логичнее быть на предыдущей строке.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,104 @@
+        setConstraints()
+        createSearchBar()
+        interactor.loadCarsList(query: "")
+        
```
Вот эта пустая строка не нужна. Не оставляй ничего висящего в конце метода. А вот предыдущий кусок я бы сгруппировал по-другому. Как минимум, у тебя там есть чисто UI-блок (всякие настройки и addSubview) и есть обращение к логике (вызов интерактора), вот их можно было бы пустой стиркой разделить.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        delegate = ProfileViewController()
```
Ты здесь допускаешь ту же ошибку, с которой приходил ко мне на неделе.

Создание какого-то вью контроллера происходит тогда и **только** тогда, когда ты хочешь открыть новый экран. Новый - т.е. его еще нет в иерархии контроллеров приложения. Тогда твои действия следующие - создать контроллер, показать его, что-то с ним делать дальше. Ты же помнишь, что если ты в коде создал вьюху или контроллер, то это еще не значит, что она должна показаться на экране? Чтобы на экране появилась вьюха, ты делаешь addSubview. Чтобы на экране появился новый вью контроллер, ты либо презентишь его, либо пушишь через навигационный стек, либо кладешь в rootViewController у UIWindow приложения. Ну или в таб бар засовываешь. В общем - каждый раз, когда ты создаешь новый контроллер, ты обязан показывать его с помощью какого-то контроллера, который уже существует на экране.

Что у тебя происходит здесь. ProfileViewController у тебя в приложении, очевидно, уже есть. Поскольку ты попал на CarsModelViewController именно через него. То есть, тебе не нужно создавать этот контроллер, он у тебя уже где-то есть. А здесь ты создаешь левый контроллер (по сути, второй профайл контроллер), не добавляешь его на экран, но все равно просишь что-то сделать. Разумеется, он не будет работать так, как тебе того хочется) 

Еще раз - новый контроллер создаем только при показе нового экрана приложения. Если экран уже есть - нам нужно работать с ссылкой на уже существующий контроллер. Следовательно, тебе нужно решить, опять же, одну из классических задач в разработке (про которую также любят спрашивать на собесах), как организовать связь между двумя модулями. Как это делается - будем обсуждать завтра. Пока просто можешь подумать, как тебе сделать так, чтобы CarsModelViewController ссылался на уже существующий ProfileViewController, т.е. как тебе сюда прокинуть ссылку на него. Дам наводку на один из возможных вариантов - в метод build() у CarsModelBuilder можно добавить новый аргумент.

YarMag replied:
Кстати, задай себе еще один вопрос. Если можно создавать контроллеры вот так вот от балды, внутри других контроллеров, то зачем нам тогда билдеры, которые как раз строго регламентируют - кто, когда, кого, как, в ответ на какие действия создает? Давай договоримся, что в рамках VIP-архитектуры, по которой мы с тобой работает, любое создание контроллера происходит только внутри его собственного билдера. Все. В любых других местах создавать контроллеры (если только это не системные штуки, типа UIAlertViewController) категорически запрещено.

И да, в рандомных местах контроллера делать MyBuilder().build() - это ровно тот же самый грех) Билдеры, в свою очередь, у нас дергаются только в роутерах (ну либо в AppDelegate, где у нас роутеров еще толком нет и мы пытаемся задать какую-то точку для входа).

YarMag commented on autoService/Profile/ProfileInfo.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    func toJson() -> [String: Any] {
+        return ["email": self.email, "phone": self.phone, "name": self.name, "city": self.city, "experience": self.experience, "birthday": self.birthday,"brand": self.brand, "model": self.model, "year": self.year, "volume": self.volume, "mileage": self.mileage] as [String: Any]
```
У тебя не вызывает дискомфорт чтение вот этого кода? Если вызывает, то абсолютно заслуженно. В линтерах на проектах обычно прям правило зашито, что длина строки не может превышать какого-то определенного числа, в противном случае проект не скомпилируется, и тогда хочешь не хочешь - придется фиксить)

 Как правило, любую длинную строку кода можно привести в человекочитаемое состояние. Например, код в init читается легко, а тут... Подумай, как можно отшлифовать.

YarMag commented on autoService/Profile/ProfileInteractor.swift:
```diff
@@ -11,11 +11,30 @@ final class ProfileInteractor {
-    let authService: AuthService
-    
-    init(authService: AuthService) {
+    init(authService: AuthService,
```
Не любишь ты марки у инициализаторов) А зря)

YarMag commented on autoService/Profile/ProfileRouter.swift:
```diff
@@ -7,13 +7,28 @@
     }
     
+    func listCarsBrand() {
+
```
Пустая строка в начале функции ничем не обоснована, не надо так делать. Вообще после открывающей скобки метода и перед закрывающей отступы никто не ставит.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        settings()
+        setupConstraints()
+        interactor.onViewDidLoad()
+        CarsModelViewController().delegate = self
```
Дружище, ну вот тут прям совсем залет. Начнем с того, что у тебя этот объект умирает  уже на 198 строке кода. Почему - на него нет ни одной сильной ссылки. 

Еще раз. Вью контроллеры - это не синглтоны. Когда ты пишешь
let controller = MyViewController()

он тебе не вернет MyViewController, который ты создавал где-то в другом месте. Он создаст полностью новый объект. Развиваем мысль - если ты напишешь
let controller = MyViewController()
let controller2 = MyViewController()

то у тебя будет два абсолютно разных вью контроллера, которые друг про друга ничего не знают. Поэтому, когда ты из CarsModel контроллера пытался дернуть ProfileController(), происходило то же самое - у тебя просто создавался еще один контроллер для профиля, про который приложение не знает. Ну а если оно про него не знает, то и работать с ним нормально ты не сможешь. Пытаться переприсвоить уже существующий контроллер, если что - ну типа попытаться закостылить, в CarsModelController создавать ProfileController, показывать его еще раз как рутовый и прикидывать нужные значения - это будет еще хуже, т.к. ты поимеешь всю свою навигацию известно куда, в обход всех правил. 

Строго говоря, CarsModelController даже не должен знать про то, какие контроллеры его используют, и кому он там данные отдает. Есть протокол - все, больше ничего не надо, он будет стучаться в кого-то, кого ему скормили под протоколом. Кстати, тут ты молодец, без протокола такую ситуацию и правда не разрулить.

YarMag commented on autoService/Profile/Cars Info/CarsModel/CarsModelViewController.swift:
```diff
@@ -0,0 +1,104 @@
+import Foundation
+import UIKit
+
+protocol CarsBrand {
```
За использование протокола - бесспорно, плюс. Однако имя мне не нравится. CarsBrand - звучит как переменная. Или сложная структура с кучей полей. Но вот протоколы так не называют.

Иногда самые простые решения - самые лучшие. В данном случае ты можешь назвать протокол CarsModelViewControllerDelegate.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        CarsModelViewController().delegate = self
+    }
+    
+    func choused(brand: String) {
```
Почему не присваивается и не работает - объяснил выше. А в имени метода опечатка, кстати, сверься с Гугл транслейтом.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+    
+    private func saveInfo() {
+        print("saveInfo  self.brand.text = \(self.brand.text!)")
+        info = ProfileInfo.init(json: ["email": email.text!, "phone": phone.text!, "name": name.text!, "city": city.text!, "experience": experience.text!, "birthday": birthday.text!, "brand": brand.text!, "model": model.text!, "year": year.text!, "volume": volume.text!, "mileage": mileage.text!])
```
А это что за красота такая) Мы ж договаривались прятать.

YarMag replied:
За читабельность километровой строки уже в принципе молчу)

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
     }
+    
+    @objc private func didTapEdit() {
+        email.isUserInteractionEnabled.toggle()
```
Хитро сделал, неплохо. Только вот копипасты многовато. Попробуй все контролы сложить в массив [UIView], а потом сделать цикл по массиву и каждого элемента сделать isUserInteractionEnabled.toggle(). Возможно, выйдет чуть более читаемо.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        info = ProfileInfo.init(json: ["email": email.text!, "phone": phone.text!, "name": name.text!, "city": city.text!, "experience": experience.text!, "birthday": birthday.text!, "brand": brand.text!, "model": model.text!, "year": year.text!, "volume": volume.text!, "mileage": mileage.text!])
+    }
+    
+    @objc private func bdayHandleDatePicker(sender: UIDatePicker) {
```
Договаривались такие вещи (bday) не сокращать)

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        year.resignFirstResponder()
+    }
+    
+    private func settings() {
```
Называть метод (т.е. действие) существительным - не фен-шуй (кстати, в стайл гайдах об этом тоже пишут). Settings() - буквально, настройки. Войдовский метод, который называется "настройки", выглядит странно. Я бы понял еще, если бы это был какой-то геттер и использовался у какого-то сложного объекта, ну типа
let settings = config.settings()

А в таком виде - не надо. Переименуй на что-то более подходящее. Даже просто setupViews() подойдет. Тут тебе и действие, и понятно, зачем оно.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        return datePicker
+    }()
+    
+    private let modelYaerDatePicker: UIDatePicker = {
```
Очепятка в имени.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+        return label
+    }()
+    
+    private let bdayDatePicker: UIDatePicker = {
```
bday не сокращаем.

YarMag commented on autoService/Profile/ProfileViewController.swift:
```diff
@@ -7,25 +7,399 @@
+            volume.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -10),
+            volume.bottomAnchor.constraint(equalTo: volume.topAnchor, constant: 50),
+            
+            mileage.topAnchor.constraint(equalTo: volume.bottomAnchor, constant: 15),
```
Помнишь, я просил отступы делать кратными 4? Потому что это стандарт. Давай не 10, 15 и 50, а хотя б 12, 16 и 48.

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -45,7 +46,25 @@ final class DBService {
+    }
+    
+    func uploadProfileInfo(profileInfo: ProfileInfo) {
+        
```
Висящая строка, не надо так.

jaha01 committed:
fixes

Вроде CarsModelViewController ссылается на уже существующий ProfileViewController
у меня 3 вопроса по твоим замечаниям, на созвоне задам

jaha01 committed:
finally done

jaha01 committed:
last

jaha01 committed:
hope it is last one
